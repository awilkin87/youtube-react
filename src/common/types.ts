
/** Information about video. Used for displaying video previews. */
export interface VideoPreviewData {
  /** Video id. */
  id: any;
  /** Title of the video. */
  videoTitle: string;
  /** Title of the channel that published the video. */
  channelTitle: string;
  /** URL to thumbnail image. */
  thumbnail: string;
  /** Number times the video has been watched. */
  viewCount: number;
  /** Video duration in ISO8601 format. */
  duration: string;
  /** Date/time the video was published. */
  publishDate: Date;
}

/** Map of video category ids to category names. */
export type VideoCategories = Record<string, string>


export const MockVideo: VideoPreviewData = {
  id: 0,
  videoTitle: 'Video Title',
  channelTitle: 'Channel Name',
  thumbnail: 'http://via.placeholder.com/210x118',
  viewCount: 56,
  duration: 'PT3M52S',
  publishDate: new Date(0)
}
