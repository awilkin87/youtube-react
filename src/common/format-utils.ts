
/** 
 * Shorten an integer value to a string with at most three digits and appending K, M, B to indicate
 * thousands, millions, or billions, respectively.
 * @param value number to shorten
 * @param locale locale for determining string format. Default locale is used if undefined.
 * @returns shortened string representation of the value
 */
export function shortenNumber(value: number, locale?: string): string {
  if (value === 0) {
    return '0';
  }

  const suffixScale = ['', 'K', 'M', 'B'];

  const digits = Math.trunc(Math.log10(value)) + 1;

  const suffixIdx = Math.trunc((digits - 1) / 3);
  const orderOfMangitude = 3 * suffixIdx;

  let significand = value / (10 ** orderOfMangitude);
  if (significand < 10) {
    // Keep up to one decimal place
    significand = Math.trunc(10 * significand) / 10;
  } else {
    // Truncate all decimal places
    significand = Math.trunc(significand);
  }

  return significand.toLocaleString(locale) + suffixScale[suffixIdx];
}
