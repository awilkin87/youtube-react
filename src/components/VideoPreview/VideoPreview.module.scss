@use '../../styles/shared.module';

.videoPreview {
  display: grid;

  &.vertical {
    grid: 118px auto / 210px; 
  }

  &.horizontal {
    grid: auto / 210px auto;
    column-gap: 4px;
  }
}

.thumbnail {
  grid-row: 1 / 2;
  grid-column: 1 / 2;

  // Make thumbnail the closest positioned ancestor for timeLabel, without removing it from the
  // document flow.
  position: relative;

  .timeLabel {
    color: white;
    font-weight: 500;
    background: black;
    opacity: 0.8;
    border-radius: 2px;

    position: absolute;
    bottom: 0;
    right: 0;
    margin: 4px;
    padding: 3px 4px;
    line-height: 12px;
  }
}

.videoInfo {
  color: shared.$secondary-text-color;
  font-size: 13px;

  a {
    color: shared.$secondary-text-color;
  }

  .videoTitle {
    color: shared.$primary-text-color;
    font-size: 14px;
    font-weight: 500;
    margin-bottom: 5px;

    // Limit to two lines; Does not work w/ IE
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 2;
    overflow: hidden;
  }

  .channelTitle {
    // Limit to one line
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
}
