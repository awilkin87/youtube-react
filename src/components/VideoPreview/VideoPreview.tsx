import { Image } from 'semantic-ui-react';
import styles from './VideoPreview.module.scss';

export interface VideoPreviewProps {
  orientation: 'horizontal' | 'vertical';
}

/** Video preview containing thumbnail image and video information. */
const VideoPreview: React.FunctionComponent<Readonly<VideoPreviewProps>> = function (props) {
  return (
    <div className={`${styles.videoPreview} ${orientationStyle(props.orientation)}`}>
      <div className={styles.thumbnail}>
        <Image src='http://via.placeholder.com/210x118' />
        <div className={styles.timeLabel}>
          <span>05:22</span>
        </div>
      </div>
      <div className={styles.videoInfo}>
        <div className={styles.videoTitle}>Video title</div>
        <div className={styles.channelTitle}>Channel title</div>
        <div>2.1M views â€¢ 2 days ago</div>
      </div>
    </div>
  );
}

function orientationStyle(orientation: VideoPreviewProps['orientation']) {
  switch (orientation) {
    case 'horizontal':
      return styles.horizontal;
    case 'vertical':
      return styles.vertical;
  }
}

export default VideoPreview;
