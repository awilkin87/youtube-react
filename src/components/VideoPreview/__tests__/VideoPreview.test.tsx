import { shallow } from "enzyme";
import { MockVideo } from "../../../common/types";
import VideoPreview from '../VideoPreview';

describe('VideoPreview', () => {
  test('Renders vertical VideoPreview', () => {
    const wrapper = shallow(<VideoPreview video={MockVideo} orientation='vertical' />);
    expect(wrapper).toMatchSnapshot();
  });

  
  test('Renders horizontal VideoPreview', () => {
    const wrapper = shallow(<VideoPreview video={MockVideo} orientation='horizontal' />);
    expect(wrapper).toMatchSnapshot();
  });
});
