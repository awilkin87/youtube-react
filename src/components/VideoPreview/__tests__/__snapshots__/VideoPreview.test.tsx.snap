// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoPreview Renders horizontal VideoPreview 1`] = `
<div
  className="videoPreview horizontal"
>
  <Link
    to="/watch?v=0"
  >
    <div
      className="thumbnail"
    >
      <Image
        as="img"
        src="http://via.placeholder.com/210x118"
        ui={true}
      />
      <div
        className="timeLabel"
      >
        <span>
          3:52
        </span>
      </div>
    </div>
  </Link>
  <div
    className="videoInfo"
  >
    <Link
      to="/watch?v=0"
    >
      <div
        className="videoTitle"
      >
        Video Title
      </div>
    </Link>
    <div
      className="channelTitle"
    >
      Channel Name
    </div>
    <Link
      to="/watch?v=0"
    >
      <div>
        56
         views • 
        <TimeAgo
          date={1970-01-01T00:00:00.000Z}
          live={false}
          title=""
        />
      </div>
    </Link>
  </div>
</div>
`;

exports[`VideoPreview Renders vertical VideoPreview 1`] = `
<div
  className="videoPreview vertical"
>
  <Link
    to="/watch?v=0"
  >
    <div
      className="thumbnail"
    >
      <Image
        as="img"
        src="http://via.placeholder.com/210x118"
        ui={true}
      />
      <div
        className="timeLabel"
      >
        <span>
          3:52
        </span>
      </div>
    </div>
  </Link>
  <div
    className="videoInfo"
  >
    <Link
      to="/watch?v=0"
    >
      <div
        className="videoTitle"
      >
        Video Title
      </div>
    </Link>
    <div
      className="channelTitle"
    >
      Channel Name
    </div>
    <Link
      to="/watch?v=0"
    >
      <div>
        56
         views • 
        <TimeAgo
          date={1970-01-01T00:00:00.000Z}
          live={false}
          title=""
        />
      </div>
    </Link>
  </div>
</div>
`;
